ext/standard/string.c

```diff
@@ -3752,6 +3752,141 @@ PHPAPI zend_string *php_str_to_str(char *haystack, size_t length, char *needle,
 }
 /* }}} */
 
+/* {{{ php_str_to_array_ex
+ */
+PHPAPI zend_string *php_str_to_array_ex(char *haystack, size_t length,
+	char *needle, size_t needle_len, HashTable *arr, int case_sensitivity, size_t *replace_count)
+{
+	zval *entry;
+	zend_string *new_str;
+	HashPosition pos;
+	size_t new_length, offset, str_len;
+	char *checked_haystack, *checked_needle, *haystack_dup = NULL
+		, *needle_dup = NULL, *end, *p, *r, *n, *str;
+	size_t max_len=0, count=0;
+
+	/* Convert replace elements to string before entering the replace loop */
+	/* and compute max length */
+
+	ZEND_HASH_FOREACH_VAL(arr, entry) {
+		SEPARATE_ZVAL(entry);
+		convert_to_string(entry);
+		if (Z_STRLEN_P(entry) > max_len) max_len = Z_STRLEN_P(entry);
+	} ZEND_HASH_FOREACH_END();
+
+	if (needle_len < length) { /* Usual case: length(needle) < length(haystack) */
+		if (case_sensitivity) {
+			checked_haystack = haystack;
+			checked_needle = needle;
+		} else {	/* Case-insensitive: lowercase check */
+			checked_haystack = haystack_dup = estrndup(haystack, length);
+			checked_needle = needle_dup = estrndup(needle, needle_len);
+			php_strtolower(haystack_dup, length);
+			php_strtolower(needle_dup, needle_len);
+		}
+
+		/* Pass 1 : compute max length of result string */
+		p = checked_haystack;
+		n = checked_needle;
+		end = p + length;
+		while ((p = (char*)php_memnstr(p, n, needle_len, end))) {
+			p += needle_len;
+			count++;
+		}
+		new_length = length + count * (max_len - needle_len);
+
+		/* Pass 2 : Replace */
+		zend_hash_internal_pointer_reset_ex(arr, &pos);
+		new_str = zend_string_alloc(new_length, 0);
+		offset=0;
+		for (p = checked_haystack, end = p+length; (r = (char*)php_memnstr(p, checked_needle, needle_len, end)); p = r + needle_len) {
+			if ((entry = zend_hash_get_current_data_ex(arr, &pos)) == NULL) {
+				if (zend_hash_num_elements(arr) == 0) { /* Empty replace array */
```

Comment:
Well, maybe, but I just made changes to support 'array search' as a real extension of 'string search'. What this means is that it allows cyclic replace when search is an array too (in this case, elements of the replace array will be strings or arrays).If an element of the replace array is an array, it is used for cyclic replace.

Do you think we should stop everything if we detect an empty array replacement ? It is probably the cleanest way to do it. I am not very familiar with PHP error handling, could the error be trapped, do we need to return a (partially) valid result, even after raising the error ?